services/wireguard/endpoint/userspace/client.go

```diff
@@ -18,27 +18,115 @@
 package userspace
 
 import (
+	"encoding/base64"
 	"errors"
 	"net"
 
+	"git.zx2c4.com/wireguard-go/device"
+	"git.zx2c4.com/wireguard-go/tun"
 	"github.com/mysteriumnetwork/node/consumer"
 	wg "github.com/mysteriumnetwork/node/services/wireguard"
 )
 
-type client struct{}
+type client struct {
+	tun    tun.TUNDevice
+	devAPI *device.DeviceApi
+}
 
 // NewWireguardClient creates new wireguard user space client.
 func NewWireguardClient() (*client, error) {
-	return nil, errors.New("Not implemented")
+	return &client{}, nil
 }
 
-func (c *client) ConfigureRoutes(iface string, ip net.IP) error { return nil }
-func (c *client) DestroyDevice(name string) error               { return nil }
-func (c *client) AddPeer(name string, peer wg.PeerInfo) error   { return nil }
-func (c *client) Close() error                                  { return nil }
-func (c *client) PeerStats() (consumer.SessionStatistics, error) {
-	return consumer.SessionStatistics{}, nil
+func (c *client) ConfigureDevice(name string, config wg.DeviceConfig, subnet net.IPNet) (err error) {
+	if c.tun, err = tun.CreateTUN(name, device.DefaultMTU); err != nil {
+		return err
+	}
+	if err := assignIP(name, subnet); err != nil {
+		return err
+	}
+
+	c.devAPI = device.UserspaceDeviceApi(c.tun)
+	if err := c.devAPI.SetListeningPort(uint16(config.ListenPort())); err != nil {
+		return err
+	}
+
+	key, err := base64stringTo32ByteArray(config.PrivateKey())
+	if err != nil {
+		return err
+	}
+
+	if err := c.devAPI.SetPrivateKey(device.NoisePrivateKey(key)); err != nil {
+		return err
+	}
+
+	c.devAPI.Boot()
```

Comment:
What this Boot performs? Can't it fail?

Reply:
It brings the device up: https://github.com/mysteriumnetwork/wireguard-go/blob/master/device/exported_device.go#L101
In the current implementation, it returns nothing.